source/simulator/solver_schemes.cc

```diff
@@ -431,12 +431,17 @@ namespace aspect
         compute_current_constraints ();
       }
 
-    // the Stokes matrix depends on the viscosity. if the viscosity
-    // depends on other solution variables, then after we need to
-    // update the Stokes matrix in every time step and so need to set
-    // the following flag. if we change the Stokes matrix we also
-    // need to update the Stokes preconditioner.
-    rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+    // If the Stokes matrix depends on the solution, or we have active
+    // velocity boundary conditions, we need to re-assemble the system matrix
+    // (and preconditioner) every time. If we have active boundary conditions,
+    // they could a) depend on the solution, or b) be inhomogeneous. In both
+    // cases, just assembling the RHS will be incorrect.  If no active
+    // boundaries exist, we only have no-slip or free slip conditions, so we
+    // don't need to force assembly of the matrix.
+    if (stokes_matrix_depends_on_solution()
+        ||
+        (boundary_velocity_manager.get_active_boundary_velocity_conditions().size() > 0))
+      rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
```

Comment:
so the following?
```
    if (stokes_matrix_depends_on_solution()
        ||
        (nonlinear_iteration == 0 && boundary_velocity_manager.get_active_boundary_velocity_conditions().size() > 0))
      rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
```

Reply:
Yes, I think that *should* work.

Reply:
ok, I will try and see what happens (I think the tester coverage should be good enough for this)

Reply:
So, the dilitation is not happy with this, but the other tests are fine. I will add a test for whether the dilation is enabled or not.

Reply:
It passes now, but since I do not really understand the dilation I hope @tjhei and/or @naliboff can take a look whether the solution is valid since they introduced it in #3426. One concern I have is that it only rebuilds the stokes matrix (which is on the rhs for the Newton solver) and not the newton stokes matrix (which is the lefthand side for the Newton solver). Is that fine, and if the test passes is that good enough for correctness?

Reply:
oke, I guess didn't wait for the tester to start. It actually failed, but this is the whole diff:

```
diff --git a/tests/prescribed_dilation_newton/screen-output b/tests/prescribed_dilation_newton/screen-output
index 03b27752f..36409e779 100644
--- a/tests/prescribed_dilation_newton/screen-output
+++ b/tests/prescribed_dilation_newton/screen-output
@@ -1,10 +1,6 @@
------------------------------------------------------------------------------
------------------------------------------------------------------------------
 
 Loading shared library <./libprescribed_dilation_newton.so>
 
------------------------------------------------------------------------------
------------------------------------------------------------------------------
 Number of active cells: 64 (on 4 levels)
 Number of degrees of freedom: 948 (578+81+289)
 
@@ -17,14 +13,13 @@ Number of degrees of freedom: 948 (578+81+289)
       Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 1, norm of the rhs: 441.548
 
    Skipping temperature solve because RHS is zero.
-   Rebuilding Stokes preconditioner...
    Solving Stokes system... 0+25 iterations.
       Relative nonlinear residual (total Newton system) after nonlinear iteration 2: 3.74883e-16, norm of the rhs: 1.65529e-13, newton_derivative_scaling_factor: 0
 
 
    Postprocessing:
      RMS, max velocity:        2.13 m/s, 4.82 m/s
-     Errors u_L2, p_L2:        1.794314e-04, 8.890806e-05
+     Errors u_L2, p_L2:        1.794314e-04, 8.890805e-05
      Writing graphical output: output-prescribed_dilation_newton/solution/solution-00000
 
 Number of active cells: 256 (on 5 levels)
@@ -39,7 +34,6 @@ Number of degrees of freedom: 3,556 (2,178+289+1,089)
       Relative nonlinear residual (total Newton system) after nonlinear iteration 1: 1, norm of the rhs: 605.137
 
    Skipping temperature solve because RHS is zero.
-   Rebuilding Stokes preconditioner...
    Solving Stokes system... 0+27 iterations.
       Relative nonlinear residual (total Newton system) after nonlinear iteration 2: 6.09492e-16, norm of the rhs: 3.68826e-13, newton_derivative_scaling_factor: 0
 
@@ -53,5 +47,3 @@ Termination requested by criterion: end time
 
 
 
------------------------------------------------------------------------------
------------------------------------------------------------------------------
```